### PR TITLE
Stabilize `destructuring_assignment`

### LINKED ISSUE #71126: Tracking Issue for `destructuring_assignment`
This is a tracking issue for the RFC "Destructuring assignment" (https://github.com/rust-lang/rfcs/pull/2909).
The feature gate for the issue is `#![feature(destructuring_assignment)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also uses as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Open a pull request: https://github.com/rust-lang/rust/pull/71156
	- [x] Tuples (https://github.com/rust-lang/rust/pull/78748)
    - [x] Structs and slices (https://github.com/rust-lang/rust/pull/78836)
    - [x] `_` expressions (https://github.com/rust-lang/rust/pull/79016)
- [x] Open an RFC: https://github.com/rust-lang/rfcs/pull/2909.
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [x] Stabilization PR (https://github.com/rust-lang/rust/pull/90521)

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Implementation history

The initial implementation was carried out in three steps by @fanzier (split out from an initial implementation, https://github.com/rust-lang/rust/pull/71156):
- Implementing destructuring support for tuples: https://github.com/rust-lang/rust/pull/78748
- Implementing destructuring support for structs and slices: https://github.com/rust-lang/rust/pull/78836
- Implementing destructuring support for `_` expressions: https://github.com/rust-lang/rust/pull/79016

### PR DESCRIPTION
**Pull request by jhpratt**
Closes #71126

- [Stabilization report](https://github.com/rust-lang/rust/issues/71126#issuecomment-941148058)
- [Completed FCP](https://github.com/rust-lang/rust/issues/71126#issuecomment-954914819)

@rustbot label +F-destructuring-assignment +T-lang
Also needs +relnotes but I don't have permission to add that tag.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
r? rust-lang/lang

**Comment by michaelwoerister** [comment]
r? rust-lang/lang-team

**Comment by joshtriplett** [comment]
This looks correct to me.

**Comment by matklad** [comment]
Do we have tests and overall plan regarding the interaction between destructing assignment and let else? With both features on track to stabilization, I feel we should figure out their interaction at least on some level *before* stabilization, least we realize we need some future-proofing somewhere *after* stabilization.

**Comment by jhpratt** [comment]
I can't speak as to the status of tests and whatnot, but I would not think there's any interaction at all. After all, let-else inherently requires `let`, which is not present for destructing assignment.

**Comment by matklad** [comment]
> After all, let-else inherently requires let, which is not present for destructing assignment.

I don’t think it inherently requires let. I can imagine the following working:

```rust
let mut x = …;
…
Some(x) = foo() else { return };
```

Anecdotally, couple of time I wanted to use let-else, I needed exactly this form, to update an existing mutable binding.

**Comment by jhpratt** [comment]
I'll try to see if I can come up with some regex to find a relevant test. If not I'll just add a test myself.

**Comment by Aloso** [comment]
The stabilization report doesn't mention destructing references. Is the plan to stabilize that later?

> ```rust
> let mut x = …;
> …
> Some(x) = foo() else { return };
> ```

@matklad This code is not accepted by the parser at the moment, but I don't see any reason why that couldn't be added later.

**Comment by jhpratt** [comment]
After a quick check, confirming that the parser doesn't accept this. I don't see any potential for compatibility concerns; it seems reasonable that this would be accepted (both syntactically and semantically) in the future.

**Comment by matklad** [comment]
I think I see at least some syntactic ambiguities of dangling-else variety:

```
let Some(x) = Some(y) = foo() else { return };
Some(x) = Some(y) = foo() else { return };
```

I don't think it makes sense to decide what are we going to do with them right now, but I do think that we should have an unresolved question somewhere, which at least outlines design space and good and poor interactions.

To clarify, the above is not an exhaustive list of potential problems, it's rather a counterexample to "it'll obviously just work together".

**Comment by veber-alex** [comment]
I wanted to check how `(a,b) = (b,a)` compares to `mem::swap`, it looks like `mem::swap` applies optimizations when `T` is 32 bytes or bigger but destructuring_assignment doesn't.

https://godbolt.org/z/czfTv3nzW

**Comment by jhpratt** [comment]
Let's keep this PR related to the stabilization and potential compatibility concerns. For optimization issues with `destructuring_assignment` it would probably be best to open a new, explicit issue.

**Comment by jhpratt** [comment]
Rebased with CI passing.

**Comment by jhpratt** [comment]
Rebased. Hopefully with a recent PR landing merge conflicts will be less likely from now on.

**Comment by jackh726** [comment]
> How do docs look currently? Are the reference and book updated?

This is still an open question from me. If all docs are updated (or PRs ready for them), then I can review this.

**Comment by jhpratt** [comment]
I have not personally looked at documentation. Not sure if anyone else has.

**Comment by jackh726** [comment]
Okay, in that case, I'm going to defer to @rust-lang/lang whether we should block this on requiring PRs to docs (or maybe just issues opened).

**Comment by varkor** [comment]
For what it's worth, I don't think there is that much that needs to be written to update the documentation, especially using the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/2909-destructuring-assignment.md) for reference. (Unfortunately, I have very little time at the moment; otherwise I would do it myself.)

File: compiler/rustc_feature/src/accepted.rs [hunk 1]
```diff
@@ -112,6 +112,8 @@ declare_features! (
     (accepted, default_type_params, "1.0.0", None, None),
     /// Allows `#[deprecated]` attribute.
     (accepted, deprecated, "1.9.0", Some(29935), None),
+    /// Allows the use of destructuring assignments.
+    (accepted, destructuring_assignment, "1.49.0", Some(71126), None),
```
Comment (jhpratt) [inline compiler/rustc_feature/src/accepted.rs:116]: Note to self: the version isn't correct here.

**Comment by jhpratt** [comment]
This is still waiting on docs to my knowledge. I will resolve the trivial conflict once that is done.

**Comment by varkor** [comment]
I've opened up a documentation pull request for the Reference (https://github.com/rust-lang/reference/pull/1116). But stabilisation is a blocker for the Reference pull request rather than the other way around. I think this is just waiting for @pnkfelix's review.

**Comment by jhpratt** [comment]
Ah, alright. I will get to fixing the conflict soon. Realistically the conflict doesn't affect a review in any way, as it's just a change from the most recent bootstrap.

**Comment by jhpratt** [comment]
Rebased.

**Comment by varkor** [comment]
Personally, I wouldn't worry about continually rebasing: `feature_gate.rs` in particular is liable to frequent conflicts. As you say, it doesn't affect the review, and you can just rebase after approval.

**Comment by jhpratt** [comment]
Yeah, I was just doing that so it _could_ be merged. Feature gates conflict much less now than they used to, as they're alphabetized instead of sorted by version. But I definitely intend to hold off until review is complete at this point.

File: src/test/ui/issues/issue-77218.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+fn main() {
```
Comment (jackh726) [inline src/test/ui/issues/issue-77218.rs:1]: What is this test for?

Should probably be moved to a more specific directory than `issues`

**Review by jackh726 — APPROVED** [review]
One question/comment. But given that there has already been a completed FCP and doc PRs are ready, r=me when comment addressed

File: src/test/ui/issues/issue-77218.rs (hunk 2 above)
Comment (jhpratt) [reply src/test/ui/issues/issue-77218.rs:1]: It's a regression test for a previous ICE in #77218. The code is taken verbatim from [this comment](https://github.com/rust-lang/rust/issues/77218#issuecomment-702107741). Where should it be placed? The issues directory is typically used for ICE regression tests, no?

Comment (jackh726) [reply src/test/ui/issues/issue-77218.rs:1]: Wait, I'm confused...this test [already exists](https://github.com/rust-lang/rust/blob/master/src/test/ui/issues/issue-77218/issue-77218.rs) and the associated issue is already closed? Why is this getting added?

Comment (jhpratt) [reply src/test/ui/issues/issue-77218.rs:1]: Whoops. Just looked at the full diff and see that this test already exists. I probably just messed something up with tab autocomplete on the command line and never noticed.

I'll get this fixed tomorrow most likely and approve it on your review at that point (I think I'm able to do that, not positive).

**Comment by jackh726** [comment]
@bors delegate+

Just r=jackh726 once you've removed the duplicate test :)

**Comment by pnkfelix** [comment]
LGTM. Thanks @jhpratt ! (And sorry for how much time I've taken to look at this.)

**Comment by jhpratt** [comment]
Thanks @pnkfelix. No worries about the delay; I was under the impression this was waiting on docs anyways.

**Comment by matthiaskrgr** [comment]
@bors r-
I think a test needs an update?
https://github.com/rust-lang/rust/pull/91897#issuecomment-993269873

**Comment by jhpratt** [comment]
Surprised that running `x t --bless` wasn't sufficient. Either way, I've done it on the clippy tests specifically. Let's give this another shot.

@bors r=jackh726,pnkfelix

**Pull request merged**